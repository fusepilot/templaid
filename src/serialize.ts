import * as fs from "fs-extra";
import { resolve, relative, basename, join, parse } from "path";

export interface IFile {
  kind: "file";
  name: string;
  path: string;
  size: number;
  content: string;
}

export interface IDirectory {
  kind: "directory";
  name: string;
  path: string;
  children: Array<IFile | IDirectory>;
}

export type Item = IFile | IDirectory;

const recurse = async ({
  path = __dirname,
  root
}: {
  path: string;
  root: string;
}): Promise<Item> => {
  path = resolve(path);

  const name = basename(path);
  let item = { path, name };

  const stat = await fs.stat(path);

  if (stat.isFile()) {
    const buffer = await fs.readFile(path);
    const content = buffer.toString();

    return {
      name,
      path: relative(root, path),
      kind: "file",
      size: stat.size,
      content
    };
  } else if (stat.isDirectory()) {
    let children = (await fs.readdir(path)) as string[];

    const items = await Promise.all(
      children.map(async child => {
        return await recurse({ path: join(path, child), root });
      })
    );

    return {
      name,
      path: relative(root, path),
      kind: "directory",
      children: items
    };
  } else {
    throw new Error(`error reading ${path}`);
  }
};

export const tree = async ({ path = __dirname }): Promise<Item> => {
  return await recurse({ path, root: path });
};
