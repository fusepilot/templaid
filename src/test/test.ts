import { renderTemplate } from '../'
import { tree } from '../serialize'
import { remove } from 'fs-extra'
import { resolve } from 'path'

it('complex', async () => {
  const outputPath = resolve(__dirname, 'output-complex')

  const result = await renderTemplate({
    templatePath: resolve(__dirname, 'templates', 'complex'),
    destinationPath: outputPath,
    data: {
      template: {
        name: 'NewProject'
      }
    }
  })

  const structure = await tree({ path: outputPath })

  expect(structure).toMatchSnapshot()

  await remove(outputPath)
})
