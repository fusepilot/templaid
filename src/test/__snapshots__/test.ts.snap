// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "content": "NewProject that should be parsed
",
          "kind": "file",
          "name": "NewProject-a.md",
          "path": "NewProject/NewProject-a.md",
          "size": 33,
        },
        Object {
          "content": "{{template.name}} that shouldnt be parsed
",
          "kind": "file",
          "name": "NewProject-b.md",
          "path": "NewProject/NewProject-b.md",
          "size": 42,
        },
        Object {
          "content": "file c content
",
          "kind": "file",
          "name": "file-c.md",
          "path": "NewProject/file-c.md",
          "size": 15,
        },
      ],
      "kind": "directory",
      "name": "NewProject",
      "path": "NewProject",
    },
    Object {
      "children": Array [
        Object {
          "content": "",
          "kind": "file",
          "name": ".hidden.json",
          "path": "folder-b/.hidden.json",
          "size": 0,
        },
        Object {
          "content": "",
          "kind": "file",
          "name": "NewProject-folder-c",
          "path": "folder-b/NewProject-folder-c",
          "size": 0,
        },
        Object {
          "children": Array [
            Object {
              "content": "{{template.name}} should be replaced. file-e
",
              "kind": "file",
              "name": "file-e.md",
              "path": "folder-b/NewProject-folder-d/file-e.md",
              "size": 45,
            },
          ],
          "kind": "directory",
          "name": "NewProject-folder-d",
          "path": "folder-b/NewProject-folder-d",
        },
        Object {
          "content": "file d content
",
          "kind": "file",
          "name": "file-d.md",
          "path": "folder-b/file-d.md",
          "size": 15,
        },
      ],
      "kind": "directory",
      "name": "folder-b",
      "path": "folder-b",
    },
  ],
  "kind": "directory",
  "name": "output-complex",
  "path": "",
}
`;
